pkg/didcomm/protocol/introduce/service.go

```diff
@@ -490,113 +410,194 @@ func isNoOp(s state) bool {
 	return ok
 }
 
-// isSkipProposal is a helper function to determine whether this is skip proposal or no
-// the usage is correct when it is executed on the introducer side
-func isSkipProposal(msg *metaData) bool {
-	// the skip proposal can be determined only after receiving dependency
-	// dependency is injected by the Continue function
-	if msg.dependency == nil {
+// isSkipProposal is a helper function to determine whether this is skip proposal or not.
+func isSkipProposal(md *metaData) bool {
+	if md.msg.Metadata()[metaSkipProposal] == nil {
 		return false
 	}
 
-	// if introducer provides an invitation this is definitely the skip proposal
-	return msg.dependency.Invitation() != nil
+	return md.msg.Metadata()[metaSkipProposal].(bool)
 }
 
-// injectInvitation the function tries to set an invitation we got from introducee
-func injectInvitation(msg *metaData) error {
-	// we can inject invitation only when we received the Response message
-	// the Response message comes always from introduce
-	if msg.Msg.Type() != ResponseMsgType {
-		return nil
+func (s *Service) handle(md *metaData) error {
+	if err := s.saveResponse(md); err != nil {
+		return err
 	}
 
-	// if we already have an invitation it has more priority
-	// the second invitation is not important for us and we are ignoring it
-	if msg.Invitation != nil {
-		return nil
-	}
+	var (
+		current   = md.state
+		actions   []stateAction
+		stateName string
+	)
 
-	// increases counter to determine from who we got an invitation
-	msg.IntroduceeIndex++
+	for !isNoOp(current) {
+		stateName = current.Name()
 
-	var resp = Response{}
+		next, action, err := s.execute(current, md)
+		if err != nil {
+			return fmt.Errorf("execute: %w", err)
+		}
 
-	if err := msg.Msg.Decode(&resp); err != nil {
-		return err
+		actions = append(actions, action)
+
+		if !isNoOp(next) && !current.CanTransitionTo(next) {
+			return fmt.Errorf("invalid state transition: %s --> %s", current.Name(), next.Name())
+		}
+
+		current = next
 	}
 
-	// sets an invitation to which will be forwarded later
-	msg.Invitation = resp.Invitation
+	if err := s.saveStateName(md.contextID, stateName); err != nil {
+		return fmt.Errorf("failed to persist state %s: %w", stateName, err)
+	}
+
+	for _, action := range actions {
+		if action == nil {
+			continue
+		}
+
+		if err := action(); err != nil {
+			return err
+		}
+	}
 
 	return nil
 }
 
-func (s *Service) handle(msg *metaData) error {
-	current := stateFromName(msg.StateName)
+func contextInvitation(msg service.DIDCommMsg) *didexchange.Invitation {
+	inv := &didexchange.Invitation{}
 
-	for !isNoOp(current) {
-		next, err := s.execute(current, msg)
-		if err != nil {
-			return fmt.Errorf("execute: %w", err)
+	switch v := msg.Metadata()[metaInvitation].(type) {
+	case service.DIDCommMsgMap:
+		if err := v.Decode(inv); err != nil {
+			// should never happen, otherwise, the protocol logic is broken
```

Comment:
OR the data model changed between versions and we didn't handle versioning properly.

Reply:
Done! I have removed panic